modules/javafx.graphics/src/main/native-glass/gtk/glass_key.cpp

```diff
@@ -31,227 +31,214 @@
+    { GDK_KEY_KP_6, com_sun_glass_events_KeyEvent_VK_NUMPAD6 },
+    { GDK_KEY_KP_7, com_sun_glass_events_KeyEvent_VK_NUMPAD7 },
+    { GDK_KEY_KP_8, com_sun_glass_events_KeyEvent_VK_NUMPAD8 },
+    { GDK_KEY_KP_9, com_sun_glass_events_KeyEvent_VK_NUMPAD9 },
+    { GDK_KEY_F1, com_sun_glass_events_KeyEvent_VK_F1 },
+    { GDK_KEY_F2, com_sun_glass_events_KeyEvent_VK_F2 },
+    { GDK_KEY_F3, com_sun_glass_events_KeyEvent_VK_F3 },
+    { GDK_KEY_F4, com_sun_glass_events_KeyEvent_VK_F4 },
+    { GDK_KEY_F5, com_sun_glass_events_KeyEvent_VK_F5 },
+    { GDK_KEY_F6, com_sun_glass_events_KeyEvent_VK_F6 },
+    { GDK_KEY_F7, com_sun_glass_events_KeyEvent_VK_F7 },
+    { GDK_KEY_F8, com_sun_glass_events_KeyEvent_VK_F8 },
+    { GDK_KEY_F9, com_sun_glass_events_KeyEvent_VK_F9 },
+    { GDK_KEY_F10, com_sun_glass_events_KeyEvent_VK_F10 },
+    { GDK_KEY_F11, com_sun_glass_events_KeyEvent_VK_F11 },
+    { GDK_KEY_F12, com_sun_glass_events_KeyEvent_VK_F12 },
+    { GDK_KEY_Shift_L, com_sun_glass_events_KeyEvent_VK_SHIFT },
+    { GDK_KEY_Shift_R, com_sun_glass_events_KeyEvent_VK_SHIFT },
+    { GDK_KEY_Control_L, com_sun_glass_events_KeyEvent_VK_CONTROL },
+    { GDK_KEY_Control_R, com_sun_glass_events_KeyEvent_VK_CONTROL },
+    { GDK_KEY_Caps_Lock, com_sun_glass_events_KeyEvent_VK_CAPS_LOCK },
+    { GDK_KEY_Meta_L, com_sun_glass_events_KeyEvent_VK_WINDOWS },
+    { GDK_KEY_Meta_R, com_sun_glass_events_KeyEvent_VK_CONTEXT_MENU },
+    { GDK_KEY_Alt_L, com_sun_glass_events_KeyEvent_VK_ALT },
+    { GDK_KEY_Alt_R, com_sun_glass_events_KeyEvent_VK_ALT_GRAPH },
+    { GDK_KEY_Super_L, com_sun_glass_events_KeyEvent_VK_WINDOWS },
+    { GDK_KEY_Super_R, com_sun_glass_events_KeyEvent_VK_WINDOWS },
+    { GDK_KEY_Delete, com_sun_glass_events_KeyEvent_VK_DELETE }
+};
+
+jint gdk_keyval_to_glass(guint keyval) {
+    int min = 0;
+    int max = G_N_ELEMENTS(gdk_to_glass_keys) - 1;
+    int mid;
+
+    // binary search
+    while (max >= min) {
+        mid = (min + max) / 2;
+        if (gdk_to_glass_keys[mid].gdk_key < keyval) {
+            min = mid + 1;
+        } else if (gdk_to_glass_keys[mid].gdk_key > keyval) {
+            max = mid - 1;
+        } else {
+            return gdk_to_glass_keys[mid].glass_key;
+        }
     }
-}
 
-jint gdk_keyval_to_glass(guint keyval)
-{
-    init_keymap();
-    return GPOINTER_TO_INT(g_hash_table_lookup(keymap, GINT_TO_POINTER(keyval)));
+    return 0;
 }
 
 jint get_glass_key(GdkEventKey* e) {
-    init_keymap();
-
     guint keyValue;
     guint state = e->state & GDK_MOD2_MASK; //NumLock test
 
     gdk_keymap_translate_keyboard_state(gdk_keymap_get_default(),
             e->hardware_keycode, static_cast<GdkModifierType>(state), e->group,
             &keyValue, NULL, NULL, NULL);
 
-    jint key = GPOINTER_TO_INT(g_hash_table_lookup(keymap,
-            GINT_TO_POINTER(keyValue)));
+    jint key = gdk_keyval_to_glass(e->keyval);
 
+    //FIXME: is this still needed?
     if (!key) {
```

Comment:
Is this still needed? I was "born" in [JDK-8115541](https://bugs.openjdk.org/browse/JDK-8115541) but Gtk3 may be handling this now.